src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,199 @@
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Diagnostics.Analyzers.NamingStyles;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Simplification;
+using static Microsoft.CodeAnalysis.Diagnostics.Analyzers.NamingStyles.SymbolSpecification;
+
+namespace Microsoft.CodeAnalysis.ConvertAutoPropertyToFullProperty
+{
+    internal abstract class AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider
+        : CodeRefactoringProvider
+    {
+        internal abstract SyntaxNode GetProperty(SyntaxToken token);
+        internal abstract string GetUniqueName(string fieldName, IPropertySymbol property);
+        internal abstract SyntaxNode GetInitializerValue(SyntaxNode property);
+        internal abstract SyntaxNode GetPropertyWithoutInitializer(SyntaxNode property);
+        internal abstract (SyntaxNode newGetAccessor, SyntaxNode newSetAccessor) GetNewAccessors(
+            DocumentOptionSet options, SyntaxNode property, string fieldName, SyntaxGenerator generator);
+        internal abstract SyntaxNode GetTypeBlock(SyntaxNode syntaxNode);
+        internal abstract SyntaxNode ConvertPropertyToExpressionBodyIfDesired(DocumentOptionSet options, SyntaxNode fullProperty);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(context.Span.Start);
+
+            var property = GetProperty(token);
+            if (property == null)
+            {
+                return;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var propertySymbol = semanticModel.GetDeclaredSymbol(property) as IPropertySymbol;
+            if (propertySymbol == null)
+            {
+                return;
+            }
+
+            if (!(IsValidAutoProperty(property, propertySymbol)))
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(
+                new ConvertAutoPropertyToFullPropertyCodeAction(
+                    FeaturesResources.Convert_to_full_property,
+                    c => ExpandToFullPropertyAsync(document, property, propertySymbol, root, cancellationToken)));
+        }
+
+        internal bool IsValidAutoProperty(SyntaxNode property, IPropertySymbol propertySymbol)
+        {
+            var fields = propertySymbol.ContainingType.GetMembers().OfType<IFieldSymbol>();
+            var field = fields.FirstOrDefault(f => propertySymbol.Equals(f.AssociatedSymbol));
```

Comment:
Don't have to change it, but `return fields.Any(f => ...)` would avoid the subsequent null check.